File: src/main/java/no/nav/familie/ks/sak/util/Ressurs.java
package no.nav.familie.ks.sak.util;

import com.fasterxml.jackson.databind.JsonNode;

public class Ressurs {
    public enum Status {
        SUKSESS, FEILET, IKKE_HENTET;
    }

    private Status status;
    private JsonNode data;
    private String melding;

    Ressurs(Status status, JsonNode data, String melding) {
        this.status = status;
        this.data = data;
        this.melding = melding;
    }

    public Status getStatus() {
        return status;
    }

    public JsonNode getData() {
        return data;
    }

    public String getMelding() {
        return melding;
    }

    public static class Builder {
        public Ressurs byggVellyketRessurs(JsonNode data) {
            return new Ressurs(Status.SUKSESS, data, "Suksess");
        }

        public Ressurs byggFeiletRessurs(String melding) {
            return new Ressurs(Status.FEILET, null, melding);
        }
    }
}

File: src/main/java/no/nav/familie/ks/sak/app/rest/Ressurs.kt
package no.nav.familie.ks.sak.app.rest

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper

data class Ressurs(
        val data: JsonNode?,
        val status: Status,
        val melding: String,
        val errorMelding: String?
) {
    enum class Status { SUKSESS, FEILET, IKKE_HENTET }

    companion object {
        val objectMapper: ObjectMapper = ObjectMapper()

        inline fun <reified T> success(data: T): Ressurs {
            return Ressurs(
                    data = objectMapper.valueToTree(data),
                    status = Status.SUKSESS,
                    melding = "Innhenting av data var vellykket",
                    errorMelding = null
            )
        }

        fun failure(errorMessage: String? = null, error: Throwable? = null): Ressurs = Ressurs(
            data = null,
            status = Status.FEILET,
            melding = errorMessage ?: "Kunne ikke hente data: ${error?.message}",
            errorMelding = error?.message
        )
    }
}
